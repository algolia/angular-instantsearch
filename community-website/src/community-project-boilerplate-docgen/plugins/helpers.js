const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
