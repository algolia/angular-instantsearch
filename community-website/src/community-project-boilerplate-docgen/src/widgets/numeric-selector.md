---
title: Widgets - Numeric Selector
layout: widget.pug
---

# Numeric Selector

## Description

The Numeric Selector component lets the user choose between numerical refinements from a dropdown menu.

## Options

| Attributes  | Type                                   | Description
| -           | -                                      | -
| `attribute` | `string`                               | Name of the numeric attribute to use
| `items`     | `{value: number, label: string}[]`     | The list of indexes to search in
| `operator?  | `"<" / "<=" / "=" / ">=" / ">" / "!="` | The operator to use to refine

## Code example

```html
<ng-ais-numeric-selector
  attribute="rating"
  operator="="
  [items]="[
    { label: 'No rating selected', value: undefined },
    { label: 'Rating: 5', value: 5 },
    { label: 'Rating: 4', value: 4 },
    { label: 'Rating: 3', value: 3 },
    { label: 'Rating: 2', value: 2 },
    { label: 'Rating: 1', value: 1 }
  ]"
>
</ng-ais-numeric-selector>
```
