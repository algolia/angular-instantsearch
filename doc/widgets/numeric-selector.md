## `<ng-ais-numeric-selector />`

The NumericSelector component lets the user choose between numerical refinements from a dropdown menu.

| Attributes  | Type                                   | Description
| -           | -                                      | -
| `attribute` | `string`                               | Name of the numeric attribute to use
| `items`     | `{value: number, label: string}[]`     | The list of indexes to search in
| `operator?  | `"<" / "<=" / "=" / ">=" / ">" / "!="` | The operator to use to refine
