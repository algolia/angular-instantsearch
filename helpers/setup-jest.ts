import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
