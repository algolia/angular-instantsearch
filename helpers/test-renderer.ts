import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
