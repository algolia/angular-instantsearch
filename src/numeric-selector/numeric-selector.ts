import { Component, Input, Inject, PLATFORM_ID } from "@angular/core";
import { connectNumericSelector } from "instantsearch.js/es/connectors";
import { noop } from "lodash-es";

import { BaseWidget } from "../base-widget";
import { NgAisInstance } from "../instantsearch/instantsearch-instance";

export type NumericSelectorState = {
  currentRefinement?: string;
  options: {}[];
  refine: Function;
};

@Component({
  selector: "ng-ais-numeric-selector",
  template: `
    <div [class]="cx('')">
      <select
        [class]="cx('select')"
        (change)="state.refine($event.target.value)"
      >
        <option
          [class]="cx('option')"
          *ngFor="let item of state.options"
          [value]="item.value"
          [selected]="item.value === state.currentRefinement"
        >
          {{item.label}}
        </option>
      </select>
    </div>
  `
})
export class NgAisNumericSelector extends BaseWidget {
  // connector options
  @Input() public attribute: string;
  @Input() public operator: "<" | "<=" | "=" | ">=" | ">" | "!=" = "=";
  @Input()
  public items: {
    value: number;
    label: string;
  }[];

  public state: NumericSelectorState = {
    currentRefinement: null,
    options: [],
    refine: noop
  };

  constructor(
    @Inject(PLATFORM_ID) public platformId: Object,
    searchInstance: NgAisInstance
  ) {
    super(searchInstance, "NumericSelector");
  }

  public ngOnInit() {
    this.createWidget(connectNumericSelector, {
      attributeName: this.attribute,
      operator: this.operator,
      options: this.items
    });
    super.ngOnInit();
  }
}
