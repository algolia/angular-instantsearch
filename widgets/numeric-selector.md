## `<ng-ais-numeric-selector />`

The NumericSelector component lets the user choose between numerical refinements from a dropdown menu.

| Attributes      | Type                                   | Description
| -               | -                                      | -
| `attributeName` | `string`                               | Name of the numeric attribute to use
| `options`       | `{value: number, label: string}[]`     | The list of indexes to search in
| `operator?`     | `"<" / "<=" / "=" / ">=" / ">" / "!="` | The operator to use to refine
| `header?`       | `string`                               | Displays text before widget
| `footer?`       | `string`                               | Displays text after widget