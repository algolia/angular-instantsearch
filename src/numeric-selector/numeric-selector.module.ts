import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";

import { NgAisNumericSelector } from "./numeric-selector";

@NgModule({
  declarations: [NgAisNumericSelector],
  entryComponents: [NgAisNumericSelector],
  exports: [NgAisNumericSelector],
  imports: [CommonModule]
})
export class NgAisNumericSelectorModule {}
