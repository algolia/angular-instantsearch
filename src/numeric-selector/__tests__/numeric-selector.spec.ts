import { createRenderer } from '../../../helpers/test-renderer';
import { NgAisNumericSelector } from '../numeric-selector';

const defaultState = {
  currentRefinement: 10,
  options: [
    { label: 'foo', value: 10 },
    { label: 'bar', value: 20 },
    { label: 'foobar', value: 30 },
    { label: 'barfoo', value: 40 },
  ],
  refine: jest.fn(),
};

const render = createRenderer({
  defaultState,
  template: '<ais-numeric-selector></ais-numeric-selector>',
  TestedWidget: NgAisNumericSelector,
});

describe('NumericSelector', () => {
  it('renders markup without state', () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it('renders markup with state', () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it('should call `refine()` on select change', () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    const select = fixture.debugElement.nativeElement.querySelector('select');
    select.value = '20';
    select.dispatchEvent(new Event('change'));

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith('20');
  });
});
