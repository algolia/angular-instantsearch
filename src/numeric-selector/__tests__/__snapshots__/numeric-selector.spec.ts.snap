// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumericSelector renders markup with state 1`] = `
<ng-component
  testedWidget={[Function NgAisNumericSelector]}
>
  <ais-instantsearch>
    <ais-numeric-selector>
      <div
        class="ais-NumericSelector"
      >
        <select
          class="ais-NumericSelector-select"
        >
          
          <option
            class="ais-NumericSelector-option"
            value="10"
          >
             foo 
          </option>
          <option
            class="ais-NumericSelector-option"
            value="20"
          >
             bar 
          </option>
          <option
            class="ais-NumericSelector-option"
            value="30"
          >
             foobar 
          </option>
          <option
            class="ais-NumericSelector-option"
            value="40"
          >
             barfoo 
          </option>
        </select>
      </div>
    </ais-numeric-selector>
  </ais-instantsearch>
</ng-component>
`;

exports[`NumericSelector renders markup without state 1`] = `
<ng-component
  testedWidget={[Function NgAisNumericSelector]}
>
  <ais-instantsearch>
    <ais-numeric-selector>
      <div
        class="ais-NumericSelector"
      >
        <select
          class="ais-NumericSelector-select"
        >
          
        </select>
      </div>
    </ais-numeric-selector>
  </ais-instantsearch>
</ng-component>
`;
